src/internals/ir/callGraph.ts

```diff
@@ -0,0 +1,193 @@
+      }
+    }
+  }
+
+  private getFunctionName(
+    func: AstFunctionDef | AstReceiver | AstContractInit,
+  ): string | undefined {
+    if (func.kind === "function_def") return func.name.text;
+    if (func.kind === "contract_init") return "contract_init";
+    if (func.kind === "receiver") return "receiver";
+    return undefined;
+  }
+
+  private processStatements(statements: AstStatement[], callerId: number) {
+    for (const stmt of statements) {
+      switch (stmt.kind) {
+        case "statement_expression":
+          this.processExpression(stmt.expression, callerId);
+          break;
+        case "statement_condition":
+          this.processStatements(stmt.trueStatements, callerId);
+          if (stmt.falseStatements) {
+            this.processStatements(stmt.falseStatements, callerId);
+          }
+          break;
+        case "statement_while":
+        case "statement_until":
+        case "statement_repeat":
+        case "statement_foreach":
+          this.processStatements(stmt.statements, callerId);
+          break;
+        case "statement_try":
+          this.processStatements(stmt.statements, callerId);
+          break;
+        case "statement_try_catch":
+          this.processStatements(stmt.statements, callerId);
+          this.processStatements(stmt.catchStatements, callerId);
+          break;
+        case "statement_let":
+        case "statement_return":
+        case "statement_assign":
+          if ("expression" in stmt && stmt.expression) {
+            this.processExpression(stmt.expression, callerId);
+          }
+          break;
+        default:
+          console.warn(`Unhandled statement type: ${(stmt as any).kind}`, stmt);
+      }
+    }
+  }
+
+  private processExpression(
+    expr: AstExpression | AstStructFieldInitializer,
+    callerId: number,
+  ) {
+    let calleeId: number | undefined;
+
+    if ("kind" in expr) {
+      if (expr.kind === "static_call") {
+        const staticCall = expr as AstStaticCall;
+        calleeId = this.findOrAddFunction(staticCall.function.text);
+      } else if (expr.kind === "method_call") {
+        const methodCall = expr as AstMethodCall;
+        calleeId = this.findOrAddFunction(methodCall.method.text);
+      }
+
+      if (calleeId !== undefined) {
+        this.addEdge(callerId, calleeId);
+      }
+
+      // Recursively process nested expressions
+      if ("args" in expr && Array.isArray(expr.args)) {
+        for (const arg of expr.args) {
+          this.processExpression(arg, callerId);
+        }
+      }
+    } else if ("value" in expr) {
+      const initializer = expr as any;
+      if (initializer.value) {
+        this.processExpression(initializer.value, callerId);
+      }
+    }
+  }
+
+  private findOrAddFunction(name: string): number {
+    const existingFunc = Array.from(this.nodeMap.values()).find(
+      (node) => node.name === name,
+    );
+    if (existingFunc) return existingFunc.astId;
```

Comment:
Why this function returns `astId`? We need the unique `idx` of the node, aren't we?